Introduce batch publisher

dwsupplee opened the pull request:
Closes: https://github.com/GoogleCloudPlatform/google-cloud-php/issues/458

This PR introduces a batch publisher, building on top of @tmatsuo's very powerful batch work. 

At this stage, it would be awesome to get some general feedback as to the approach taken (and thoughts on the caveats below).

/cc @bshaffer @chingor13 @jdpedrie @mdietz94 @michaelbausor @tmatsuo

#### Example Usage
Starting up the daemon:

Please note, this is optional - if you choose to not use the daemon the batched messages will be published at the end of the request lifecycle via `register_shutdown_function()`:
```sh
export IS_BATCH_DAEMON_RUNNING=true
./src/Core/bin/google-cloud-batch daemon
```
Sample script:
```php
use Google\Cloud\PubSub\PubSubClient;

$pubsub = new PubSubClient();
$pubsub->topic('the_topic_of_your_dreams')
    ->batchPublisher()
    ->publish([
        'data' => 'What a great message!'
    ]);
```

#### Notes on the implementation

- A [suggested approach](https://github.com/GoogleCloudPlatform/google-cloud-php/issues/458#issuecomment-299614919) was add a `publishAsync` method to `Topic`. I ran into a few issues with this:

    - Passing in options related to the batch job on each call does not seem ideal. Encapsulating the options inside of a class helps to make them more reusable. It would be possible to pass the batch options along through the `Topic` constructor, but there are a number of ways users can get a `Topic` (`PubSubClient::createTopic()`, `PubSubClient::topic()`, `PubSubClient::topics()`, `Snapshot::topic()`) and it seemed cleaner to me to have one place where you supply the options.
    - Down the line, we plan to add `Async` variations of methods that trigger network requests across the veneer libraries (https://github.com/GoogleCloudPlatform/google-cloud-php/issues/31). If possible, we would like to reserve the name `publishAsync` for the use case where a user does not require batching but would like the benefit of concurrent network requests.
- The use of the `BatchPublishContainer` could be mimicked in `Google\Cloud\Logging\PsrLogger`, allowing us to not worry about needing to serialize the entire `PsrLogger` for the daemon to process incoming data. This could open us up to having a `shouldBatch` (or similar) option on the `PsrLogger`, so we only have one class users need to worry about interfacing with. This approach also has the benefit of reducing the public API surface we need to maintain in the `PsrBatchLogger` (as we could remove the sendEntries public method). Please let me know what you think @tmatsuo.
- Attempting to process the jobs in a cron job wasn't efficient as the minimum timeframe a cron job can be scheduled for is 1 minute. If anyone has any further experience with this though, I'm all ears!

#### Caveats
The daemon relies on [shared memory](http://php.net/manual/en/book.sem.php) and [PCNTL](http://php.net/manual/en/book.pcntl.php), extensions which can't always be guaranteed to be available. The latter has no windows support. Additionally, shared memory won't scale well across multiple nodes.

With that said, I do feel like this will work great for the good share of our use cases.

@mdietz94 Do you think these restrictions are acceptable for us moving forward with beta? I think it would be worth investigating a more robust message queue to handle batch processing - but that will take us more time. We could provide methods to interface with redis, mysql, etc., and look at including a daemon that is built in something a bit more portable than we have currently.

Thanks for any thoughts!

Todos
- [x] - Tests
- [x] - Ability to flush queued messages on demand

tmatsuo commented:
> The use of the BatchPublishContainer could be mimicked in Google\Cloud\Logging\PsrLogger, allowing us to not worry about needing to serialize the entire PsrLogger for the daemon to process incoming data. This could open us up to having a shouldBatch (or similar) option on the PsrLogger, so we only have one class users need to worry about interfacing with. This approach also has the benefit of reducing the public API surface we need to maintain in the PsrBatchLogger (as we could remove the sendEntries public method). Please let me know what you think @tmatsuo.

I think that means introducing something like `BatchLoggerContainer`, and think it as though it's a private class, and the `@internal` annotation makes it so. Sounds reasonable to me.

tmatsuo commented on src/PubSub/BatchPublishContainer.php:
```diff
@@ -0,0 +1,95 @@
+        }
+        $end = microtime(true);
+        if ($this->debugOutput) {
+            printf(
```
Re: these printf calls

It might be better to write to STDERR because if you don't run the daemon, those output will be on the end user's web browser.

Yeah, the psrBatchLogger has the same issue. I would fix it there too.

dwsupplee commented:
If there is anything regarding this approach we would like to discuss further please let me know, otherwise I'm going to finish up the rest of the implementation/tests now.

mdietz94 commented:
SGTM

On Mon, Jun 19, 2017 at 9:26 AM, David Supplee <notifications@github.com>
wrote:

> If there is anything regarding this approach we would like to discuss
> further please let me know, otherwise I'm going to finish up the rest of
> the implementation/tests now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/google-cloud-php/pull/534#issuecomment-309438754>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABvXaltvdAY1I1ET0VERUHtMXQd1pqVxks5sFnb8gaJpZM4N32c->
> .
>

dwsupplee commented:
All set and ready for review! If this looks good to everyone, this will allow us to move forward with getting the batch branch in master and published in our next release 😄.

A few changes worth noting:

- @tmatsuo, I removed the `PsrBatchLogger` in favor of a single `PsrLogger` class which will support both batch and sync calls (like we had discussed, please let me know if that has changed).
- Originally I had introduced the concept of `Containers` that could be serialized, but dropped that in favor of implementing the `Serializable` interface when needed instead.
- @chingor13 I updated the `AsyncReporter` with the changes I applied to logging and pubsub as well, please take a look when you have a moment to let me know if you're happy with the direction.
- It is now possible to `flush` any items from the queue prior to `register_shutdown_function` being triggered. This is intended mainly for the in memory queue and the primary use cases for this will be users who are writing their code inside of an event loop.

tmatsuo commented on src/Core/Batch/SysvTrait.php:
```diff
@@ -64,6 +64,11 @@ private function isSysvIPCLoaded()
-        return getenv('IS_BATCH_DAEMON_RUNNING') !== false;
+        $isDaemonRunning = filter_var(
+            getenv('IS_BATCH_DAEMON_RUNNING'),
+            FILTER_VALIDATE_BOOLEAN
```
:+1:

chingor13 commented:
The Trace AsyncReporter is working well in my [test app](https://github.com/chingor13/php-test-apps/tree/master/silex/batch).

michaelbausor commented:
I have done a pass, LGTM!

tmatsuo commented:
The performance of the e2e app is noticeably worse then the previous version. It's noticeable when you conduct a simple ab load test. The requests/second dropped from 269.92 to 134.32.

I think it is because the app handler needs to instantiate the LoggingClient which means it instantiate the internal Logger, where actually the Logger is not needed for the frontend request if we use the daemon.

The load test with the new version:
```
$ ab -n 10000 -c 20 https://tmatsuo-test.appspot.com/
This is ApacheBench, Version 2.3 <$Revision: 1604373 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking tmatsuo-test.appspot.com (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:        nginx
Server Hostname:        tmatsuo-test.appspot.com
Server Port:            443
SSL/TLS Protocol:       TLSv1.2,ECDHE-RSA-AES128-GCM-SHA256,2048,128

Document Path:          /
Document Length:        51 bytes

Concurrency Level:      20
Time taken for tests:   74.451 seconds
Complete requests:      10000
Failed requests:        5406
   (Connect: 0, Receive: 0, Length: 5406, Exceptions: 0)
Non-2xx responses:      1
Total transferred:      12633200 bytes
HTML transferred:       512079 bytes
Requests per second:    134.32 [#/sec] (mean)
Time per request:       148.903 [ms] (mean)
Time per request:       7.445 [ms] (mean, across all concurrent requests)
Transfer rate:          165.71 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        4    8   3.1      8      83
Processing:    50  140  45.6    130     732
Waiting:       50  140  45.6    129     732
Total:         57  149  45.9    138     741

Percentage of the requests served within a certain time (ms)
  50%    138
  66%    157
  75%    170
  80%    180
  90%    207
  95%    233
  98%    269
  99%    297
 100%    741 (longest request)
```

The load test with the old version
```
$ ab -n 10000 -c 20 https://logging-batch-old-dot-tmatsuo-test.appspot.com/
This is ApacheBench, Version 2.3 <$Revision: 1604373 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking logging-batch-old-dot-tmatsuo-test.appspot.com (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:        nginx
Server Hostname:        logging-batch-old-dot-tmatsuo-test.appspot.com
Server Port:            443
SSL/TLS Protocol:       TLSv1.2,ECDHE-RSA-AES128-GCM-SHA256,2048,128

Document Path:          /
Document Length:        50 bytes

Concurrency Level:      20
Time taken for tests:   37.047 seconds
Complete requests:      10000
Failed requests:        9962
   (Connect: 0, Receive: 0, Length: 9962, Exceptions: 0)
Total transferred:      12718268 bytes
HTML transferred:       517093 bytes
Requests per second:    269.92 [#/sec] (mean)
Time per request:       74.095 [ms] (mean)
Time per request:       3.705 [ms] (mean, across all concurrent requests)
Transfer rate:          335.25 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        4    8   2.6      8      77
Processing:    44   65  37.2     60     717
Waiting:       44   65  37.2     59     717
Total:         49   74  37.7     68     726

Percentage of the requests served within a certain time (ms)
  50%     68
  66%     71
  75%     74
  80%     75
  90%     83
  95%    103
  98%    164
  99%    191
 100%    726 (longest request)
```

The code for the new version:
```
<?php

require_once __DIR__ . '/../vendor/autoload.php';

use Google\Cloud\Logging\LoggingClient;

$client = new LoggingClient(
    ['projectId' => getenv('GCLOUD_PROJECT')]
);

$batchLogger = $client->psrLogger(
    'app',
    ['debugOutput' => false,
     'batchEnabled' => true,
     'batchOptions' => ['workerNum' => 10]
    ]
);

$num = isset($_GET['num']) ? $_GET['num'] : 1;
$size = isset($_GET['size']) ? $_GET['size'] : 1;

$start = microtime(true);

$message = str_repeat('x', $size);
for ($i = 0; $i < $num; $i++) {
    $batchLogger->info($message);
}
$end = microtime(true);
printf(
    '%d logs sent in %f seconds, at %f LPS' . PHP_EOL,
    $num, $end - $start,
    $num / ($end - $start)
);
```

The code for the old version:
```
<?php

require_once __DIR__ . '/../vendor/autoload.php';

use Google\Cloud\Logging\PsrBatchLogger;

$batchLogger = new PsrBatchLogger(
    'app',
    ['debugOutput' => false,
     'batchOptions' => ['workerNum' => 10],
     'clientConfig' => ['projectId' => getenv('GCLOUD_PROJECT')]
    ]
);

$num = isset($_GET['num']) ? $_GET['num'] : 1;
$size = isset($_GET['size']) ? $_GET['size'] : 1;

$start = microtime(true);

$message = str_repeat('x', $size);
for ($i = 0; $i < $num; $i++) {
    $batchLogger->info($message);
}
$end = microtime(true);
printf(
    '%d logs sent in %f seconds, at %f LPS' . PHP_EOL,
    $num, $end - $start,
    $num / ($end - $start)
);
```

tmatsuo commented:
One possible idea of preventing the performance drop might be to make LoggingClient and the PsrLogger Lazily create the connection.

Unfortunately, it seems to me that, that change needs to have backward incompatible change, as long as we merge PsrLogger and PsrBatchLogger.

Our options are:

1. Make them lazily create the connection so that it'll have better performance, but we'll need to bump the major version of google/cloud and google/cloud-logging.

2. Have PsrLogger and PsrBatchLogger separate for now. Merge them when we bump the major version later.

3. Embrace the performance drop.

I prefer the option 1 over others. I'd push back the option 3.

mdietz94 commented:
Just to be clear, this is a performance drop for users not using a daemon
correct? I agree with preferring option 1 and pushing back on option 3.

tmatsuo commented:
@mdietz94 This is a performance drop for users using the daemon. I don't think it affects users not using the daemon. Also, the drop is just for the LoggingClient.

mdietz94 commented:
Meaning this won't affect CP/S users? I would have expected a very large
increase in CP/S throughput/QPS with this change since it was synchronous
before.

tmatsuo commented:
@mdietz94 

CP/S might have the same trivial issue, where the PubSubClient creates the internal Connection object when unnecessary, especially when the user is using the daemon.

But, yeah as you said, the new PubSubClient will have much much better performance compared to the previous state for some cases, even without using the daemon.

dwsupplee commented:
Ouch! Yeah I don't like option 3 at all either. I toyed around a little bit and was able to get average  timings for instantiating the `LoggingClient` down from `0.003524065s` to `0.001411096s` without the need for a breaking change.

@tmatsuo Would you be able to remove [these lines](https://github.com/GoogleCloudPlatform/google-cloud-php/blob/batch/src/Logging/Connection/Rest.php#L47-L51)

and add the following

```php
        if (!$this->requestWrapper) {
            $this->setRequestWrapper(new RequestWrapper($this->config));
        }

        if (!$this->requestBuilder) {
            $this->setRequestBuilder(new RequestBuilder(
                $this->config['serviceDefinitionPath'],
                self::BASE_URI
            ));
        }
```

[here](https://github.com/GoogleCloudPlatform/google-cloud-php/blob/batch/src/Core/RestTrait.php#L75) to see if that improves the numbers a bit in another run?

dwsupplee commented:
@mdietz94 

A user will be able to directly instantiate the `BatchPublisher` class and use that without needing the overhead of creating the PubSubClient (which is why we're seeing a performance drop here). Either way they choose, however, they're still going to be seeing massive performance gains.

tmatsuo commented:
@dwsupplee Yeah, it's little bit better
```
$ ab -n 10000 -c 20 https://tmatsuo-test.appspot.com/
This is ApacheBench, Version 2.3 <$Revision: 1604373 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking tmatsuo-test.appspot.com (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:        nginx
Server Hostname:        tmatsuo-test.appspot.com
Server Port:            443
SSL/TLS Protocol:       TLSv1.2,ECDHE-RSA-AES128-GCM-SHA256,2048,128

Document Path:          /
Document Length:        50 bytes

Concurrency Level:      20
Time taken for tests:   66.856 seconds
Complete requests:      10000
Failed requests:        9816
   (Connect: 0, Receive: 0, Length: 9816, Exceptions: 0)
Total transferred:      12637531 bytes
HTML transferred:       516196 bytes
Requests per second:    149.57 [#/sec] (mean)
Time per request:       133.713 [ms] (mean)
Time per request:       6.686 [ms] (mean, across all concurrent requests)
Transfer rate:          184.59 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        4    9   3.0      9      78
Processing:    46  124 137.3     89    1080
Waiting:       46  124 137.2     89    1080
Total:         55  133 137.4     98    1089

Percentage of the requests served within a certain time (ms)
  50%     98
  66%    108
  75%    115
  80%    122
  90%    167
  95%    286
  98%    784
  99%    869
 100%   1089 (longest request)
```

dwsupplee commented:
I'm getting just about the exact same performance instantiating a `LoggingClient` with the changes I proposed just a moment ago vs. commenting out everything in the `LoggingClient` constructor.

Given that, if we want the absolute optimal performance while using the daemon, I think we might need to go back to having the `PsrBatchLogger` be a separate class. Even though it isn't used in batch mode, the overhead of creating the `Logger` looks like our bottleneck.

mdietz94 commented:
Okay great, SGTM, thanks.

tmatsuo commented:
> I would imagine most would think to use the clients to gain access to the batch objects.
Yeah, I don't disagree.

I took another look at the both code and noticed that, the PsrBatchLogger also creates the Logger object before sending the entry via the message queue. So I think I was wrong.

tmatsuo commented:
I think the performance drop came from the indirection of the `BatchTrait` and `getCallback()` which might be considered as a tax to pay for applying this to other services...

tmatsuo commented:
OK, LGTM!

tmatsuo commented on src/Core/Batch/BatchTrait.php:
```diff
@@ -0,0 +1,195 @@
+            : new BatchRunner();
+        $this->batchRunner->registerJob(
+            $this->identifier,
+            [$this, 'send'],
```
Having the common `send` function is good for maintenance, but additional function call adds certain cost within the infinite loop in the daemon.

Does it make sense to directly register the function, and remove the `send` and `getCallback` functions? There will be some copy-paste code here and there, but if it gains performance, it's gonna be a good maintenance tax to pay for the better performance.

tmatsuo replied:
Ok, I tested it, removed the `send` and `getCallback`, and directly registered the function. Unfortunately I couldn't achieve any significant performance gain.

Probably we can move forward with the current implementation and I will invest some of my time on the performance improvement in the future. Also, the end2end testing that I'm conducting is actually easily affected by the Logging API's server status, so the method is not very scientific.

The minimum bound of the performance we're getting with the current code is acceptable, so maybe we can move forward?

dwsupplee replied:
I just finished up a little testing of my own using the examples provided above. Both sets of code returned very similar performance results for me (I tested the latest code removing `send` and `getCallback` as well, and saw no difference). I'll try to run some more in a little bit to see if these results were an anomaly.

> The minimum bound of the performance we're getting with the current code is acceptable, so maybe we can move forward?

I would hate to introduce something in this PR that diminishes performance - but I do think these benchmarks are an extreme use case and most users won't be affected too harshly. I would be okay with moving forward if you are as well!

Using an older version ([this commit](https://github.com/GoogleCloudPlatform/google-cloud-php/commit/81319933b4c914d5b4bb00c88712c07437c73822)):
```
$ ab -n 10000 -c 20 localhost:8123/ex.php
This is ApacheBench, Version 2.3 <$Revision: 1706008 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:
Server Hostname:        localhost
Server Port:            8123

Document Path:          /ex.php
Document Length:        51 bytes

Concurrency Level:      20
Time taken for tests:   176.947 seconds
Complete requests:      10000
Failed requests:        1233
   (Connect: 0, Receive: 0, Length: 1233, Exceptions: 0)
Total transferred:      1618765 bytes
HTML transferred:       508765 bytes
Requests per second:    56.51 [#/sec] (mean)
Time per request:       353.893 [ms] (mean)
Time per request:       17.695 [ms] (mean, across all concurrent requests)
Transfer rate:          8.93 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.1      0       4
Processing:    20  353  39.3    354    1128
Waiting:       19  352  39.3    353    1127
Total:         20  353  39.3    354    1128

Percentage of the requests served within a certain time (ms)
  50%    354
  66%    359
  75%    361
  80%    363
  90%    368
  95%    373
  98%    378
  99%    385
 100%   1128 (longest request)
```

And the code introduced in this PR ([commit](https://github.com/GoogleCloudPlatform/google-cloud-php/pull/534/commits/5a9c89d5a1005092017860ab89d6dd0468ee2b19)):
```
$ ab -n 10000 -c 20 localhost:8123/ex.php
This is ApacheBench, Version 2.3 <$Revision: 1706008 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking localhost (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:
Server Hostname:        localhost
Server Port:            8123

Document Path:          /ex.php
Document Length:        51 bytes

Concurrency Level:      20
Time taken for tests:   176.260 seconds
Complete requests:      10000
Failed requests:        7717
   (Connect: 0, Receive: 0, Length: 7717, Exceptions: 0)
Total transferred:      1627711 bytes
HTML transferred:       517711 bytes
Requests per second:    56.73 [#/sec] (mean)
Time per request:       352.520 [ms] (mean)
Time per request:       17.626 [ms] (mean, across all concurrent requests)
Transfer rate:          9.02 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        0    0   0.1      0       4
Processing:    20  352  21.8    352     669
Waiting:       19  351  21.8    351     668
Total:         20  352  21.8    352     669

Percentage of the requests served within a certain time (ms)
  50%    352
  66%    358
  75%    361
  80%    363
  90%    368
  95%    372
  98%    378
  99%    382
 100%    669 (longest request)
```

tmatsuo replied:
Sure, go ahead and merge. Nice work!

tmatsuo commented:
Then now I understand why I'm getting worse performance.

I tested the app with the ErrorReporting Bootstrap enabled. ErrorReporting then create a `LoggingClient` without specifying the `projectId`. It will issue a request to the metadata server for getting the project id.

tmatsuo commented:
Yes, I get little better result
```
$ ab -n 10000 -c 20 https://tmatsuo-test.appspot.com/
This is ApacheBench, Version 2.3 <$Revision: 1604373 $>
Copyright 1996 Adam Twiss, Zeus Technology Ltd, http://www.zeustech.net/
Licensed to The Apache Software Foundation, http://www.apache.org/

Benchmarking tmatsuo-test.appspot.com (be patient)
Completed 1000 requests
Completed 2000 requests
Completed 3000 requests
Completed 4000 requests
Completed 5000 requests
Completed 6000 requests
Completed 7000 requests
Completed 8000 requests
Completed 9000 requests
Completed 10000 requests
Finished 10000 requests


Server Software:        nginx
Server Hostname:        tmatsuo-test.appspot.com
Server Port:            443
SSL/TLS Protocol:       TLSv1.2,ECDHE-RSA-AES128-GCM-SHA256,2048,128

Document Path:          /
Document Length:        51 bytes

Concurrency Level:      20
Time taken for tests:   65.239 seconds
Complete requests:      10000
Failed requests:        7371
   (Connect: 0, Receive: 0, Length: 7371, Exceptions: 0)
Total transferred:      12638241 bytes
HTML transferred:       516432 bytes
Requests per second:    153.28 [#/sec] (mean)
Time per request:       130.477 [ms] (mean)
Time per request:       6.524 [ms] (mean, across all concurrent requests)
Transfer rate:          189.18 [Kbytes/sec] received

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:        4    9   3.0      9      85
Processing:    49  121  36.2    116     452
Waiting:       49  120  36.1    115     451
Total:         62  130  36.5    125     458

Percentage of the requests served within a certain time (ms)
  50%    125
  66%    139
  75%    148
  80%    154
  90%    175
  95%    191
  98%    216
  99%    245
 100%    458 (longest request)
```

with having a file named `ErrorReportingBootstrap` in the project root.

```
<?php

require __DIR__ . '/vendor/autoload.php';

use Google\Cloud\Logging\LoggingClient;

return (new LoggingClient(['projectId' => getenv('GCLOUD_PROJECT')]))
    ->psrLogger('app-error', [
        'batchEnabled' => true,
        'debugOutput' => true,
        'batchOptions' => [
            'workerNum' => 2
        ]
    ]);
```

Does it make sense to read this environment variable before accessing Metadata server? This must affect other Client instantiation too.

tmatsuo commented:
https://docs.google.com/document/d/126NJTDNctB0zL8dwTo8-9indyWGUXpGkrXCKrHU4b-s/edit#heading=h.8t69iwn5s7dr

Actually it broke the symphony integration in the above doc.
Does it make sense to revive the `PsrBatchLogger` for easier dependency injection?

dwsupplee commented:
Good find, that explains the differences between our test runs as I was running my code locally.

> Does it make sense to read this environment variable before accessing Metadata server? This must affect other Client instantiation too.

Yes! That sounds like a great idea.

> Actually it broke the symphony integration in the above doc.
Does it make sense to revive the PsrBatchLogger for easier dependency injection?

Would [configuring a factory](https://symfony.com/doc/current/service_container/factories.html) help solve this problem?

tmatsuo commented:
Created #551
